### ISSUE TITLE
The speed of iteration of distributed training is slower than single instance's.

### ISSUE THREAD

**Issue by Farrellow** [bytedance/byteps#67]
**Describe the bug**
I ran both single-node-training and distributed-training respectively. The speed of iteration of single-training is about 100 samples/sec, and the speed of iteration of distributed-training is about 50 samples/sec per node. The single-node-training is in one node with GPU, and the distributed-training is in two nodes with 1GPU/node(there are altogether 2GPUs). the sum of two nodes speed of iteration is about 100 samples/sec(the same as single-node-training), but it cost two times of resource as single-node-training. Why???

**To Reproduce**
1. Building Docker images
I used Docker to run tasks, here is my Dockerfile: 
[Dockerfile.server](https://github.com/bytedance/byteps/files/3405289/Dockerfile.server.txt), [Dockerfile.worker.mxnet.cu90](https://github.com/bytedance/byteps/files/3405295/Dockerfile.worker.mxnet.cu90.txt)
2. I have 4 nodes to run distributed training with byteps, 2 workers, 2 servers, and 1 scheduler. All of the nodes' speed of network is 20000Mb/s(`ethtool bond0 | grep Speed`). 
192.168.0.101: worker-0
192.168.0.104: worker-1
192.168.0.56: server-0, scheduler
192.168.0.7: server-1
3. Run distributed training
192.168.0.101 worker-0:
```
export DMLC_ROLE=worker;
export DMLC_PS_ROOT_URI=192.168.0.56;
export DMLC_PS_ROOT_PORT=9000;
export DMLC_WORKER_ID=0;
export DMLC_NUM_WORKER=2;
export DMLC_NUM_SERVER=2;
python /usr/local/byteps/launcher/launch.py python train_bps.py --network r100 --loss arcface --dataset retina --models-root ./models/r100-5x5 --device 0 --per-batch-size 60
```
192.168.0.104 worker-1:
```
export DMLC_ROLE=worker;
export DMLC_PS_ROOT_URI=192.168.0.56;
export DMLC_PS_ROOT_PORT=9000;
export DMLC_WORKER_ID=1;
export DMLC_NUM_WORKER=2;
export DMLC_NUM_SERVER=2;
python /usr/local/byteps/launcher/launch.py python train_bps.py --network r100 --loss arcface --dataset retina --models-root ./models/r100-5x5 --device 0 --per-batch-size 60
```
192.168.0.56 server-0:
```
export DMLC_ROLE=server;
export DMLC_PS_ROOT_URI=192.168.0.56;
export DMLC_PS_ROOT_PORT=9000;
export DMLC_NUM_WORKER=2;
export DMLC_NUM_SERVER=2;
python /usr/local/byteps/launcher/launch.py
```
192.168.0.7 server-1:
```
export DMLC_ROLE=server;
export DMLC_PS_ROOT_URI=192.168.0.56;
export DMLC_PS_ROOT_PORT=9000;
export DMLC_NUM_WORKER=2;
export DMLC_NUM_SERVER=2;
python /usr/local/byteps/launcher/launch.py
```
192.168.0.56 scheduler:
```
export DMLC_ROLE=scheduler;
export DMLC_PS_ROOT_URI=192.168.0.56;
export DMLC_PS_ROOT_PORT=9000;
export DMLC_NUM_WORKER=2;
export DMLC_NUM_SERVER=2;
python /usr/local/byteps/launcher/launch.py
```
5. distributed log:
[worker-0.log](https://github.com/bytedance/byteps/files/3405440/worker-0.log)
[worker-1.log](https://github.com/bytedance/byteps/files/3405441/worker-1.log)
[server-0.log](https://github.com/bytedance/byteps/files/3405444/server-0.log)
[server-1.log](https://github.com/bytedance/byteps/files/3405445/server-1.log)
[scheduler.log](https://github.com/bytedance/byteps/files/3405448/scheduler.log)

6. run single-node-training:
```
python train.py --network r100 --loss arcface --dataset retina --models-root ./models/r100-5x5 --device 0 --per-batch-size 60
```
7. single-node-training log:
[single-node-training.log](https://github.com/bytedance/byteps/files/3405424/single-node-training.log)

**Expected behavior**

**Screenshots**

**Environment (please complete the following information):**
 - OS: Ubuntu 16.04.3 LTS
 - GCC version: 4.9
 - CUDA and NCCL version: 9.0 and d7a58cfa5865c4f627a128c3238cc72502649881
 - Framework (TF, PyTorch, MXNet): MXNet

**Additional context**

**Comment by bobzhuyb** [comment]
It depends on your training job -- what's the model, what's the batch size, etc. There is no guarantee that BytePS (and any other distributed motheds) can give you linear speed-up..

**Comment by Farrellow** [comment]
@bobzhuyb Yes, I know that any other DL distributed training methods can not give us linear speed-up, sometimes 55%~70% of linear speed is acceptable when training with two workers. but 50% is the same as single-node-training.

**Comment by ymjiang** [comment]
@Farrellow We need more details. Mind share the `train.py` you use? Did you use BytePS APIs to modify your code?

**Comment by Farrellow** [comment]
OK.
`train.py`:
[train.py](https://github.com/bytedance/byteps/files/3406144/train.py.txt)
`train_bps.py`:
[train_bps.py](https://github.com/bytedance/byteps/files/3406145/train_bps.py.txt)

**Comment by ymjiang** [comment]
@Farrellow Can you please try `export MXNET_OMP_MAX_THREADS=10` for the servers, and then rerun to see the performance?

**Comment by bobzhuyb** [comment]
@Farrellow 

There are a few things you can try. If any of the following works for you, please let us know. Though the following env may start with `MXNET`, they apply to any workers, TF/MXNet/PyTorch, because the parameter server is based on MXNet.

1. For the parameter servers, set `export MXNET_OMP_MAX_THREADS=10` if you have 16 CPU cores per server. Set `export MXNET_OMP_MAX_THREADS=4` if you only have 8 CPU cores

2. Set `export MXNET_CPU_WORKER_NTHREADS=32` . This may speed up the parameter server

3. Start more parameter server instances. For example, when you have two physical machines to run the servers, you can start 4 (`DMLC_NUM_SERVER=4`), two server instances per physical machine. This will increase the network bandwidth utilization especially when your single TCP flow cannot saturate your bandwidth.

**Comment by Farrellow** [comment]
> @Farrellow Can you please try `export MXNET_OMP_MAX_THREADS=10` for the servers, and then rerun to see the performance?

Thank you, I will try it.

**Comment by Farrellow** [comment]
> @Farrellow
> 
> There are a few things you can try. If any of the following works for you, please let us know. Though the following env may start with `MXNET`, they apply to any workers, TF/MXNet/PyTorch, because the parameter server is based on MXNet.
> 
> 1. For the parameter servers, set `export MXNET_OMP_MAX_THREADS=10` if you have 16 CPU cores per server. Set `export MXNET_OMP_MAX_THREADS=4` if you only have 8 CPU cores
> 2. Set `export MXNET_CPU_WORKER_NTHREADS=32` . This may speed up the parameter server
> 3. Start more parameter server instances. For example, when you have two physical machines to run the servers, you can start 4 (`DMLC_NUM_SERVER=4`), two server instances per physical machine. This will increase the network bandwidth utilization especially when your single TCP flow cannot saturate your bandwidth.

Yes, I have tried your the 3rd opinion yesterday. I found that the network speed(use `iftop -i bond0`) of node which ran server is about 5Gbps, but my physical machine's network speed is 20000Mb/s, so I thought that the bottleneck is TCP connection, and I set two servers per node, and then the speed of iteration was about 63samples/sec. It seems that distributed-training makes training speed-up a little.

I will try you other two opinions later. Thank you very much!!!

**Comment by bobzhuyb** [comment]
@Farrellow Okay. Good to know that. You can further increase the number of server instances per node... 

In this case, you don't need to try a large `MXNET_OMP_MAX_THREADS` because the bottleneck is not the CPU. Please make sure MXNET_OMP_MAX_THREADS*N is smaller than the total number of cores of your physical machine, where N is the number of server instances per physical machine. Otherwise, the servers will compete with each other.

**Comment by Farrellow** [comment]
@bobzhuyb  Thank You! My physical machine have 40 CPU cores, I set `export MXNET_OMP_MAX_THREADS=10` with 4 servers per node, and the speed of iteration is about 77 samples/sec. And I set `export MXNET_OMP_MAX_THREADS=8` with 4 servers per node, and the speed of iteration is about 81 samples/sec. The env `MXNET_OMP_MAX_THREADS` is very useful.

I still have a question.
> 2. Set `export MXNET_CPU_WORKER_NTHREADS=32` . This may speed up the parameter server
The env `MXNET_CPU_WORKER_NTHREADS`, where should I set? server or worker?

**Comment by ymjiang** [comment]
Glad that `MXNET_OMP_MAX_THREADS` works for you, probably because there are some background processes that take up too much CPU on your servers.

From our experience, `MXNET_CPU_WORKER_NTHREADS` should be able to accelerate both workers and servers for a little bit.

**Comment by Farrellow** [comment]
@bobzhuyb @ymjiang I have a new problem. I ran single-node-training with 2GPUs, `python train.py --network r100 --loss arcface --dataset retina --models-root ./models/r100-5x5 --device 0 --per-batch-size 60`, the speed of iteration is about 193 samples/sec. I ran distributed-training with 4nodes, which 2 nodes for workers, there is one worker in each node and 2 GPUs per worker, and 2nodes for server, there are 4 servers in each node. The speed of iteration is about 102 samples/sec per worker. 
```
192.168.0.101: worker-0
192.168.0.104: worker-1
192.168.0.56: scheduler
192.168.0.7: server-0, server-2, server-4, server-6
192.168.0.78: server-1, server-3, server-5, server-7
```
Changing env `MXNET_OMP_MAX_THREADS` and `MXNET_CPU_WORKER_NTHREADS` or increasing the number of servers can improve the speed, but the best performance is about 112 samples/sec per worker till this moment.

I also tried to run 4 workers on 2 nodes and 12 servers on 3 nodes, which each worker with 1 GPU, the best performance I got is about 65 samples/sec per worker till this moment.
```
192.168.0.101: worker-0, worker-2
192.168.0.104: worker-1, worker-3
192.168.0.56: scheduler
192.168.0.7: server-0, server-3, server-6, server-9
192.168.0.78: server-1, server-4, server-7, server-10
192.168.0.15: server-2, server-5, server-8, server-11
```

Is there any problem with my operation?

**Comment by ymjiang** [comment]
@Farrellow Here are [some env](https://github.com/bytedance/byteps/blob/master/docs/env.md#performance-tuning) that you can try to tune. For example, you can set `BYTEPS_PARTITION_BYTES=2048000` (default is 4096000).

Besides, have you tried Horovod? What is the performance on Horovod? Migrating the code from BytePS to Horovod should be very easy.

**Comment by Farrellow** [comment]
OK, I will try these env.
I have already tried horovod in my project, the performance was not ideal either, so I try BytePS.
